### ISSUE TITLE
bug: disabling/enabling blocks not consitent

### ISSUE THREAD

**Issue by mrkprdo** [google/blockly#7113]
### Check for duplicates

- [x] I have searched for similar issues before opening a new one.

### Description

if one of the inner block is disabled, and then you disabled few blocks on the top.
saving the json/xml and loading it again, if you enable the top level block it does not get enabled, but the context menu has already changed to "disable block", you have to disable and enable it again to work with the block.

### Reproduction steps

1. Load this json in the playground, this is a simple if-condition block
 ```json
{
  "blocks": {
    "languageVersion": 0,
    "blocks": [
      {
        "type": "controls_if",
        "id": "wGMY60,6:n?|G}f|zdt{",
        "x": 16,
        "y": 8,
        "enabled": false,
        "inputs": {
          "IF0": {
            "block": {
              "type": "logic_compare",
              "id": "d,5g$byf^3xi`z9`gv6*",
              "fields": {
                "OP": "EQ"
              },
              "inputs": {
                "A": {
                  "block": {
                    "type": "variables_get",
                    "id": "M4KwCD]HsT.Y76^8h8*0",
                    "fields": {
                      "VAR": {
                        "id": "guRlyqX$KE+(,eMeqbmp"
                      }
                    }
                  }
                },
                "B": {
                  "block": {
                    "type": "text",
                    "id": "e^gV0l)OCAFsN9}~)ok_",
                    "fields": {
                      "TEXT": "Hello World"
                    }
                  }
                }
              }
            }
          },
          "DO0": {
            "block": {
              "type": "text_print",
              "id": "U^hRVv/ZefKP:%UmLwAr",
              "inputs": {
                "TEXT": {
                  "shadow": {
                    "type": "text",
                    "id": "ce%/Ftbh0o!0Yj-Qu;]l",
                    "fields": {
                      "TEXT": "abc"
                    }
                  },
                  "block": {
                    "type": "variables_get",
                    "id": "a~/}=W]C``GDHr(Y=_J1",
                    "enabled": false,
                    "fields": {
                      "VAR": {
                        "id": "guRlyqX$KE+(,eMeqbmp"
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    ]
  },
  "variables": [
    {
      "name": "asd",
      "id": "guRlyqX$KE+(,eMeqbmp"
    }
  ]
}
```
2.  Enable the top level block which is the if-block.
3.  It does not get enabled, but context menu has changed.
4.  ~The only thing that has been done here, is the getter block for `asd` is disabled before disabling the entire blocks.~
5. Regardless of inner block is changed, as long as you re-load the json definition and re-enable the block, it is still reproducible.

### Screenshots

![image](https://github.com/google/blockly/assets/15664355/d775e3fa-700b-4ab1-a204-8f6f602130d1)

**Comment by mrkprdo** [comment]
Hi @maribethb,
After upgrading to v10.2.2, we still have this issue on our app, however trying to reproduce this on the playground again:
both here:
https://blockly-demo.appspot.com/static/tests/playgrounds/advanced_playground.html#debugEnabled=false
and
https://blockly-demo.appspot.com/static/tests/playground.html
They seem to be working... is the playground using the latest version too (v10.2.2 as of this writing)?

**Comment by maribethb** [comment]
I can only reproduce this in XML, not in JSON. The advanced playground uses XML for its save capabilities so that's why you can reproduce it there but not by loading JSON in the regular simple playground. I wrote a test case for xml equivalent to the json test case I wrote in #7172 and can confirm it fails. Guess I should have written that at the time too.

I confirmed this bug was also present in XML in Blockly v9.3.1 so I'm not sure why I thought it wasn't when I fixed this for XML. I did notice it wasn't present in XML in 9.1.1 however.

@mrkprdo are you serializing in json or xml in your app? You provided json for the reproduction steps above, but if you're still experiencing this in your app, is it with xml? I can't repro with JSON at all right now.

**Comment by mrkprdo** [comment]
@maribethb that's right, i am serializing in xml.

**Comment by maribethb** [comment]
When you call `domToWorkspace` (xml)

1. calls `domToBlockInternal`
2. The top block and its children are created
3. Those blocks get `initSvg` called on them <- disabled rendering is done here during normal rendering. importantly `visuallyDisabled` is not set. only the value of `block.isEnabled()` is checked.
4. `updateDisabled` is called **on the top block only**
5. if the top block is *not* disabled, we still call `updateDisable` on the **next** blocks but not on the **child** blocks
6. Return early, so children blocks never get their visually disabled status set if their parent is not disabled

other note: the fix in #7172 was effectively removed in #7308 so investigating that. adding it back to the `renderEfficiently` method makes the test pass

still need to look into json to see why it works.
